src/System.Resources.Binary.Reader/src/System/Resources/Binary/BinaryResourceReader.cs

```diff
@@ -0,0 +1,133 @@
+                        }
+                        break;
+                    }
+                case SerializationFormat.TypeConverterByteArray:
+                    {
+                        // read length
+                        int length = reader.Read7BitEncodedInt();
+                        if (length < 0)
+                        {
+                            throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourceDataLengthInvalid, length));
+                        }
+
+                        byte[] data = reader.ReadBytes(length);
+
+                        TypeConverter converter = TypeDescriptor.GetConverter(type);
+
+                        if (converter == null)
+                        {
+                            throw new TypeLoadException(SR.Format(SR.TypeLoadException_CannotLoadConverter, type));
+                        }
+
+                        value = converter.ConvertFrom(data);
+                        break;
+                    }
+                case SerializationFormat.TypeConverterString:
+                    {
+                        string stringData = reader.ReadString();
+
+                        TypeConverter converter = TypeDescriptor.GetConverter(type);
+
+                        if (converter == null)
+                        {
+                            throw new TypeLoadException(SR.Format(SR.TypeLoadException_CannotLoadConverter, type));
+                        }
+
+                        value = converter.ConvertFrom(stringData);
+                        break;
+                    }
+                case SerializationFormat.Stream:
+                    {
+                        // read length
+                        int length = reader.Read7BitEncodedInt();
+                        if (length < 0)
+                        {
+                            throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResourceDataLengthInvalid, length));
+                        }
+                        Stream stream;
+
+                        if (reader.BaseStream is UnmanagedMemoryStream ums)
+                        {
+                            // For the case that we've memory mapped in the .resources
+                            // file, just return a Stream pointing to that block of memory.
+                            unsafe
+                            {
+                                stream = new UnmanagedMemoryStream(ums.PositionPointer, length, length, FileAccess.Read);
+                            }
+                        }
+                        else
+                        {
+
+                            byte[] bytes = reader.ReadBytes(length);
+                            // Lifetime of memory == lifetime of this stream.
+                            stream = new MemoryStream(bytes, false);
+                        }
+
+                        value = Activator.CreateInstance(type, new object[] { stream });
```

Comment:
This could be as dangerous as `BinaryFormatter`. I think we said this was intended to operate over trusted data, correct?

Reply:
Correct: resources are assumed to be as trusted as executable code.  Even before we get here ResourceManager had the opportunity to do something similar based on the resources header (the extension point we're using here): https://github.com/dotnet/coreclr/blob/8252365bfb7651a5473e98d3bdaeaa8cb22c0ec3/src/System.Private.CoreLib/shared/System/Resources/ManifestBasedResourceGroveler.cs#L255

The point of this change is not to make resources more trusted, its to enable non-primitive resources without relying on BinaryFormatter.